### PR TITLE
update to pydantic-core 2.15.0

### PR DESCRIPTION
**Pull request by davidhewitt**
## Change Summary

Pulls in `pydantic-core` 2.15.0 with all accompanying improvements.

## Related issue number

N/A

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: pydantic/_internal/_generate_schema.py [hunk 1]
```diff
@@ -1311,22 +1309,22 @@ def _tuple_schema(self, tuple_type: Any) -> core_schema.CoreSchema:
...
                 # special case for `tuple[()]` which means `tuple[]` - an empty tuple
-                return core_schema.tuple_positional_schema([])
+                return core_schema.tuple_schema([])
         elif params[-1] is Ellipsis:
             if len(params) == 2:
-                return self._tuple_variable_schema(tuple_type, params[0])
+                return core_schema.tuple_schema([self.generate_schema(params[0])], variadic_item_index=0)
             else:
                 # TODO: something like https://github.com/pydantic/pydantic/issues/5952
                 raise ValueError('Variable tuples can only have one type')
```
Comment (davidhewitt) [inline pydantic/_internal/_generate_schema.py:1321]: @dmontagu I suspect that this is where the new variadic tuple functionality can be extended as per #5952. I'll defer this to a future PR in your hands.

File: pydantic/json_schema.py [hunk 2]
```diff
@@ -803,7 +803,9 @@ def list_schema(self, schema: core_schema.ListSchema) -> JsonSchemaValue:
         self.update_with_validations(json_schema, schema, self.ValidationsMapping.array)
         return json_schema
 
-    def tuple_positional_schema(self, schema: core_schema.TuplePositionalSchema) -> JsonSchemaValue:
+    @deprecated('tuple_positional_schema is deprecated. Use tuple_schema instead.')
+    @final
+    def tuple_positional_schema(self, schema: core_schema.TupleSchema) -> JsonSchemaValue:
```
Comment (sydney-runkle) [inline pydantic/json_schema.py:808]: I think perhaps we want to use a `deprecated` pattern similar to that enforced in this recent PR: #8294.

Specifically, here's an example: https://github.com/pydantic/pydantic/pull/8294/files#diff-2d54a873f0e89e361f635a68aff651e9507fe82cce1881943cafba93295315ae

@Viicos, thanks for the guidance on proper usage of `deprecated` going forward 🚀 😄.

File: pydantic/json_schema.py [hunk 3]
```diff
@@ -803,7 +803,9 @@ def list_schema(self, schema: core_schema.ListSchema) -> JsonSchemaValue:
...
         """Generates a JSON schema that matches a positional tuple schema e.g. `Tuple[int, str, bool]`.
```
Comment (sydney-runkle) [inline pydantic/json_schema.py:809]: Do we want to change this docstring at all based on the changes to the logic of the function? Is it necessarily **positional**?

File: pydantic/json_schema.py [hunk 4]
```diff
@@ -812,19 +814,11 @@ def tuple_positional_schema(self, schema: core_schema.TuplePositionalSchema) ->
...
-        else:
-            json_schema['maxItems'] = len(schema['items_schema'])
-        self.update_with_validations(json_schema, schema, self.ValidationsMapping.array)
-        return json_schema
+        return self.tuple_schema(schema)
 
-    def tuple_variable_schema(self, schema: core_schema.TupleVariableSchema) -> JsonSchemaValue:
+    @deprecated('tuple_variable_schema is deprecated. Use tuple_schema instead.')
+    @final
+    def tuple_variable_schema(self, schema: core_schema.TupleSchema) -> JsonSchemaValue:
```
Comment (sydney-runkle) [inline pydantic/json_schema.py:821]: Same as above re deprecated pattern

File: tests/test_json_schema.py [hunk 5]
```diff
@@ -2651,8 +2651,8 @@ def test_tuple_with_extra_schema():
     class MyTuple(Tuple[int, str]):
         @classmethod
         def __get_pydantic_core_schema__(cls, _source_type: Any, handler: GetCoreSchemaHandler) -> CoreSchema:
-            return core_schema.tuple_positional_schema(
-                [core_schema.int_schema(), core_schema.str_schema()], extras_schema=core_schema.int_schema()
+            return core_schema.tuple_schema(
+                [core_schema.int_schema(), core_schema.str_schema(), core_schema.int_schema()], variadic_item_index=2
```
Comment (sydney-runkle) [inline tests/test_json_schema.py:2655]: Why do we have the 3rd entry here as a duplicate of `core_schema.int_schema()`?

**Review by sydney-runkle — COMMENTED** [review]
Left a few questions + comments, plus some minor nits. Looking forward to discussing more 👍.

File: tests/test_validators.py [hunk 6]
```diff
@@ -63,6 +63,7 @@ class Model(BaseModel):
     assert Model(x='1.0').x == 1.0
 
 
+@pytest.mark.xfail(sys.version_info >= (3, 9) and sys.implementation == 'pypy', reason='PyPy 3.9+ bug')
```
Comment (davidhewitt) [inline tests/test_validators.py:66]: I attempted to reproduce locally but could not; I suggest to move forward we xfail this one for now, can revisit another time.

File: pydantic/_internal/_generate_schema.py [hunk 7]
```diff
@@ -180,6 +180,11 @@ def apply_each_item_validators(
     if schema['type'] == 'nullable':
         schema['schema'] = apply_each_item_validators(schema['schema'], each_item_validators, field_name)
         return schema
+    elif schema['type'] == 'tuple':
+        if (variadic_item_index := schema.get('variadic_item_index')) is not None:
```
[the remark this answers is not in the document]
Comment (davidhewitt) [reply pydantic/_internal/_generate_schema.py:184]: Haha when I first used 3.8 it was 🐳

😂

File: pydantic/json_schema.py (hunk 3 above)
Comment (davidhewitt) [reply pydantic/json_schema.py:809]: Will just axe the docstring for simplicity.

File: tests/test_json_schema.py (hunk 5 above)
Comment (davidhewitt) [reply tests/test_json_schema.py:2655]: The new `tuple_schema` doesn't have `extras_schema`, instead I add it back as the 3rd entry and make it variadic with `variadic_item_index`.

File: docs/errors/validation_errors.md [hunk 8]
```diff
@@ -372,7 +372,6 @@ This error is also raised for strict fields when the input value is not an insta
 
 This error is raised when the input value is a string that cannot be parsed for a `datetime` field:
 
-<!-- uncomment with new pydantic-core version
 ```py test="skip"
```
Comment (sydney-runkle) [inline docs/errors/validation_errors.md:375]: I think? When I was putting this in, I think `pytest-examples` was still trying to run the test even though it was commented, but I can't remember.

```suggestion
```py
```

File: pydantic/json_schema.py [hunk 9]
```diff
@@ -803,8 +805,31 @@ def list_schema(self, schema: core_schema.ListSchema) -> JsonSchemaValue:
...
-        """Generates a JSON schema that matches a positional tuple schema e.g. `Tuple[int, str, bool]`.
+    @deprecated('tuple_positional_schema is deprecated. Use tuple_schema instead.', category=None)
+    @final
+    def tuple_positional_schema(self, schema: core_schema.TupleSchema) -> JsonSchemaValue:
+        """Replaced by `tuple_schema`."""
+        warnings.warn(
+            'tuple_positional_schema is deprecated. Use tuple_schema instead.',
+            PydanticDeprecatedSince26,
+            stacklevel=2,
+        )
```
Comment (sydney-runkle) [inline pydantic/json_schema.py:816]:
```suggestion
    @deprecated('`tuple_positional_schema` is deprecated. Use `tuple_schema` instead.', category=None)
    @final
    def tuple_positional_schema(self, schema: core_schema.TupleSchema) -> JsonSchemaValue:
        """Replaced by `tuple_schema`."""
        warnings.warn(
            '`tuple_positional_schema` is deprecated. Use `tuple_schema` instead.',
            PydanticDeprecatedSince26,
            stacklevel=2,
        )
```

Comment (sydney-runkle) [reply pydantic/json_schema.py:816]: 🤷‍♀️ tiny nit

File: pydantic/json_schema.py [hunk 10]
```diff
@@ -803,8 +805,31 @@ def list_schema(self, schema: core_schema.ListSchema) -> JsonSchemaValue:
...
+
+    @deprecated('tuple_variable_schema is deprecated. Use tuple_schema instead.', category=None)
+    @final
+    def tuple_variable_schema(self, schema: core_schema.TupleSchema) -> JsonSchemaValue:
+        """Replaced by `tuple_schema`."""
+        warnings.warn(
+            'tuple_variable_schema is deprecated. Use tuple_schema instead.',
+            PydanticDeprecatedSince26,
+            stacklevel=2,
+        )
```
Comment (sydney-runkle) [inline pydantic/json_schema.py:827]:
```suggestion
    @deprecated('`tuple_variable_schema` is deprecated. Use `tuple_schema` instead.', category=None)
    @final
    def tuple_variable_schema(self, schema: core_schema.TupleSchema) -> JsonSchemaValue:
        """Replaced by `tuple_schema`."""
        warnings.warn(
            '`tuple_variable_schema` is deprecated. Use `tuple_schema` instead.',
            PydanticDeprecatedSince26,
            stacklevel=2,
        )
```

File: tests/test_json_schema.py (hunk 5 above)
Comment (sydney-runkle) [reply tests/test_json_schema.py:2655]: Great, thanks for the explanation!

File: tests/test_validators.py (hunk 6 above)
Comment (sydney-runkle) [reply tests/test_validators.py:66]: Sounds good. This one has been funky for me as well.

**Review by sydney-runkle — APPROVED** [review]
Looks great overall - a few tiny change requests, but I went ahead + approved, given that they're mostly preferential.

File: tests/test_validators.py (hunk 6 above)
Comment (sydney-runkle) [reply tests/test_validators.py:66]: Looks like this test still ran for some reason, I wonder if it was the `sys.implementation` check that didn't work...

Comment (davidhewitt) [reply tests/test_validators.py:66]: Ah yes it's `sys.implementation.name`.

**Pull request merged**